[Feature] Restrict receiving files

KaterinaWire opened the pull request:
## What's new in this PR?

Display placeholder instead of the file (audio, video, image, file) UI element and disable a possibility to download the current file when team admin/owner has disabled file sharing from the team settings. 

### Notes

https://wearezeta.atlassian.net/browse/SQSERVICES-613

### Attachments

https://zpl.io/2jy9YwW

johnxnguyen commented on Wire-iOS/Resources/Base.lproj/Localizable.strings:
```diff
@@ -1644,6 +1644,10 @@
 "feature.flag.update.file_sharing.alert.title" = "File sharing restrictions";
 "feature.flag.update.file_sharing.alert.message.disabled" = "Sharing and receiving files of any type is disabled";
 "feature.flag.update.file_sharing.alert.message.enabled" = "Sharing and receiving files of any type is enabled";
+"feature.flag.restriction.file" = "Receiving files restricted";
+"feature.flag.restriction.picture" = "Receiving pictures restricted";
+"feature.flag.restriction.video" = "Receiving videos restricted";
+"feature.flag.restriction.audio" = "Receiving audio files restricted";
```
I feel like maybe a key like `"feature_config.file_sharing_restrictions.file"` might be more suitable, what do you think?

KaterinaWire replied:
Sounds fine for me

johnxnguyen commented on Wire-iOS/Sources/Helpers/syncengine/ZMMessage+Actions.swift:
```diff
@@ -35,6 +35,12 @@ extension ZMConversationMessage {
     }
 
+    /// Whether the message can be received.
+    var canBeReceived: Bool {
+        guard let session = ZMUserSession.shared() else { return true }
+        return session.fileSharingFeature.status == .enabled
+    }
```
Does this only consider asset messages? Or also normal messages?

KaterinaWire replied:
This feature flag only applies to assets and I thought it might be more general in the future, but maybe it makes sense to rename it and have a more specific property.
`canAssetsBeReceived`, what do you think?

johnxnguyen replied:
I would also consider a name like `isRestricted` because it's simpler and is inline with other names you've selected.

But I was actually referring to whether this condition needs to check if the message is actually an asset, because we're extending `ZMConversationMessage` and I think it applies to all messages.

KaterinaWire replied:
Updates

johnxnguyen replied:
Did this change happen in another framework?

KaterinaWire replied:
Yes
https://github.com/wireapp/wire-ios-data-model/pull/1191

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionAudioCell.swift:
```diff
@@ -25,6 +25,7 @@ import WireCommonComponents
 
 final class CollectionAudioCell: CollectionCell {
     private let audioMessageView = AudioMessageView()
+    private let restrictionView = AudioMessageRestrictionView()
```
It looks like the `restrictionView` is laid out exactly like the `audioMessageView`. Can we instead just insert one or the other in the view hierarchy, instead of both?

johnxnguyen replied:
Any thoughts about this?

johnxnguyen reviewed (commented):
There's quite a bit of duplication across all the new views. Did you try to factor this all into a single view type with some configurability?

johnxnguyen commented on Wire-iOS Tests/ConversationMessageCell/ConversationImageMessageTests.swift:
```diff
@@ -24,6 +24,12 @@ final class ConversationImageMessageTests: XCTestCase {
+
+        UIColor.setAccentOverride(.vividRed)
+    }
+    
```
Lint issue here.

johnxnguyen commented on Wire-iOS Tests/Mocks/MockMessage.swift:
```diff
@@ -348,6 +348,10 @@ class MockMessage: NSObject, ZMConversationMessage, ConversationCompositeMessage
     var linkAttachments: [LinkAttachment]?
     var needsLinkAttachmentsUpdate: Bool = false
     var isSilenced: Bool = false
+    var isRestrictedBacking: Bool = false
```
To be more consistent with other properties here, I suggest to rename this to: `backingIsRestritcted`.

KaterinaWire replied:
Fixed

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationVideoMessageCell.swift:
```diff
@@ -56,37 +57,54 @@ final class ConversationVideoMessageCell: RoundedView, ConversationMessageCell {
+        transferView.isHidden = object.isObfuscated || object.message.isRestricted
+
+        if isRestricted {
+            
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/FileTransfer/FileTransferView.swift:
```diff
@@ -155,6 +155,21 @@ final class FileTransferView: UIView, TransferView {
+        self.bottomLabel.accessibilityValue = self.bottomLabel.attributedText?.string ?? ""
+
+        guard !message.isRestricted else {
+            fileEyeView.setTemplateIcon(.block, size: 8)
+
+            let firstLine = fileNameAttributed
+            let secondLine = L10n.Localizable.FeatureConfig.FileSharingRestrictions.file.localizedUppercase && labelFont && labelTextBlendedColor
+            self.topLabel.attributedText = firstLine
+            self.bottomLabel.attributedText = secondLine
+
+            self.actionButton.isUserInteractionEnabled = false
+            return
```
You could extract this into a separate method to `configureRestrictedView`, to make this more readable. Also, is `self` needed?

KaterinaWire replied:
Removed this code

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/AudioMessageRestrictionView.swift:
```diff
@@ -0,0 +1,57 @@
+    init() {
+        super.init(context: .audio)
+    }
+    
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/BaseMessageRestrictionView.swift:
```diff
@@ -0,0 +1,175 @@
+        return imageView
+    }()
+
+    private var context: MessageRestrictionContext {
```
The name `context` is has a very wide range of meanings / interpretations. Also, I typically think of an `NSManagedObjectContext` when I see the name `context`. I suggest to use a more precise name, such as `messageType`. Similarly, you could rename the type to `RestrictedMessageType`.

KaterinaWire replied:
Renamed

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/BaseMessageRestrictionView.swift:
```diff
@@ -0,0 +1,175 @@
+        setupIconView()
+
+        var allViews: [UIView] {
+            switch context {
+            case .audio:
+                return [topLabel, bottomLabel, iconView]
+            case .video, .image:
+                return [topLabel, iconView]
```
I wouldn't expect any specialized logic to be in the base class. Shouldn't we move this to the subclasses?

KaterinaWire replied:
Yes.
Moved

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/BaseMessageRestrictionView.swift:
```diff
@@ -0,0 +1,175 @@
+        iconView.contentMode = .center
+        iconView.accessibilityIdentifier = "\(context.rawValue.capitalizingFirstLetter()) + MessageRestrictionIcon"
+
+        switch context {
+        case .video:
+            iconView.clipsToBounds = true
+            iconView.layer.cornerRadius = 16
+            iconView.backgroundColor = .white
+        default:
+            break
+        }
```
Also here, this seems more appropriately placed in the subclass for the video messages.

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/BaseMessageRestrictionView.swift:
```diff
@@ -0,0 +1,175 @@
+
+}
+
+
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/BaseMessageRestrictionView.swift:
```diff
@@ -0,0 +1,175 @@
+}
+
+
+enum MessageRestrictionContext: String {
+    case audio
+    case video
+    case image
```
Don't we need a case for file messages too?

KaterinaWire replied:
Yes. I forgot to move the code for file messages into a separate class.
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/ImageMessageRestrictionView.swift:
```diff
@@ -0,0 +1,51 @@
+
+            // top label
+            topLabel.centerXAnchor.constraint(equalTo: self.centerXAnchor),
+            topLabel.topAnchor.constraint(equalTo: iconView.bottomAnchor, constant: 12),
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/VideoMessageRestrictionView.swift:
```diff
@@ -0,0 +1,51 @@
+
+    override func createConstraints() {
+        super.createConstraints()
+        
```
Lint issue here and below.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/MessagePresenter.swift:
```diff
@@ -164,11 +164,11 @@ final class MessagePresenter: NSObject {
         if Message.isLocation(message) {
             openLocationMessage(message)
-        } else if Message.isVideo(message) {
+        } else if Message.isVideo(message), !message.isRestricted  {
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/ImageMessageRestrictionView.swift:
```diff
@@ -0,0 +1,72 @@
+    // MARK: - Properties
+
+    /// For the search screen
+    private var isShortVersion: Bool
```
This flag makes the view two different views depending on its value. I suggest to keep things simple and just declare a separate view for each use case.

KaterinaWire replied:
Yes, this flag makes the view different (hide the label). But I decided to keep it here because we are using the same views as for the conversation and for the search screen. And only for image and video views, we have to hide the title.

KaterinaWire replied:
But it might be better to create separate views for them.

David-Henner replied:
If you'd like to keep this flag in the view, I would suggest naming it something more explicit

KaterinaWire replied:
I created another view and removed this flag.

David-Henner commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/Restrictions/FileMessageRestrictionView.swift:
```diff
@@ -0,0 +1,95 @@
+    func configure(for message: ZMConversationMessage?) {
+        super.configure()
+
+        guard let fileMessageData = message?.fileMessageData else { return }
+        let filepath = (fileMessageData.filename ?? "") as NSString
+        let filename = (filepath.lastPathComponent as NSString).deletingPathExtension
```
this might be better in an extension of `ZMConversationMessage`

KaterinaWire replied:
Done

David-Henner commented on Wire-iOS/Sources/UserInterface/Conversation/InputBar/ConversationInputBarViewController.swift:
```diff
@@ -468,6 +472,10 @@ final class ConversationInputBarViewController: UIViewController,
     }
 
+    @objc func updateInputBarButtons() {
+
```
will this be implemented in an other PR?

KaterinaWire replied:
Yes

David-Henner reviewed (commented):
There are some `SwiftLint` warnings :)

KaterinaWire commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationImageMessageCell.swift:
```diff
@@ -82,17 +83,20 @@ final class ConversationImageMessageCell: UIView,
 
         obfuscationView.fitInSuperview()
         imageResourceView.fitInSuperview()
+        restrictionView.fitInSuperview()
```
Removed `fitInSuperview`

KaterinaWire commented:
@johnxnguyen I can't find your comment, but I added changes to the collection cells. I added the contained and add a subview depending on the message "type".

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionAudioCell.swift:
```diff
@@ -41,54 +41,60 @@ final class CollectionAudioCell: CollectionCell {
+        view.layer.cornerRadius = 4
+        view.clipsToBounds = true
+
+        containerView.addSubview(view)
```
You might need to empty the container view before adding the subview, otherwise it may be that both views are added.

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionFileCell.swift:
```diff
@@ -55,37 +58,45 @@ final class CollectionFileCell: CollectionCell {
-        }
+        headerView.translatesAutoresizingMaskIntoConstraints = false
+        containerView.translatesAutoresizingMaskIntoConstraints = false
+        
```
Lint issue here.

KaterinaWire replied:
You don't need to point to the Lint issues, I see them tool :)

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionFileCell.swift:
```diff
@@ -55,37 +58,45 @@ final class CollectionFileCell: CollectionCell {
+        view.layer.cornerRadius = 4
+        view.clipsToBounds = true
+
+        containerView.addSubview(view)
```
Same here regarding the comment about emptying the container view.

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionImageCell.swift:
```diff
@@ -90,27 +83,36 @@ final class CollectionImageCell: CollectionCell {
+    private func setup(_ view: UIView) {
+        view.clipsToBounds = true
+
+        containerView.addSubview(view)
```
And here too.

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionVideoCell.swift:
```diff
@@ -44,31 +44,46 @@ final class CollectionVideoCell: CollectionCell {
 
     func loadView() {
+        containerView.translatesAutoresizingMaskIntoConstraints = false
+        secureContentsView.addSubview(containerView)
```
And here.

KaterinaWire replied:
Done

KaterinaWire replied:
I'm adding container here

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Collections/CollectionVideoCell.swift:
```diff
@@ -44,31 +44,46 @@ final class CollectionVideoCell: CollectionCell {
 
+    private func setup(_ view: UIView) {
+        view.clipsToBounds = true
+        containerView.addSubview(view)
```
Maybe also here?

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationAudioMessageCell.swift:
```diff
@@ -40,66 +41,58 @@ class ConversationAudioMessageCell: RoundedView, ConversationMessageCell {
-        obfuscationView.isHidden = !object.isObfuscated
-        transferView.isHidden = object.isObfuscated || object.message.isRestricted
+    private func setup(_ view: UIView) {
+        containerView.addSubview(view)
```
And here.

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationImageMessageCell.swift:
```diff
@@ -135,36 +116,51 @@ final class ConversationImageMessageCell: UIView,
 
-    func updateImageContainerAppearance() {
+    private func setup(_ view: UIView) {
+        containerView.addSubview(view)
```
Same here

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationVideoMessageCell.swift:
```diff
@@ -40,69 +41,62 @@ final class ConversationVideoMessageCell: RoundedView, ConversationMessageCell {
     }
 
+    private func setup(_ view: UIView) {
+        containerView.addSubview(view)
```
Here too

KaterinaWire replied:
Done

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Rich Content/ConversationVideoMessageCell.swift:
```diff
@@ -40,69 +41,62 @@ final class ConversationVideoMessageCell: RoundedView, ConversationMessageCell {
+            view.bottomAnchor.constraint(equalTo: containerView.bottomAnchor)
+        ])
+    }
+    
```
Lint issue here.

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Conversation/Content/Cells/ConfigurationMessageCell/Content/Text/ConversationFileMessageCell.swift:
```diff
@@ -39,67 +40,59 @@ class ConversationFileMessageCell: RoundedView, ConversationMessageCell {
-        obfuscationView.isHidden = !object.isObfuscated
-        fileTransferView.isHidden = object.isObfuscated || object.message.isRestricted
+    private func setup(_ view: UIView) {
+        containerView.addSubview(view)
```
Here too

KaterinaWire replied:
Done
